lib/collection/src/shards/replica_set/shard_transfer.rs

```diff
@@ -155,25 +132,74 @@ impl ShardReplicaSet {
+                let _ = local_write.insert(Shard::Local(local_shard));
+                Ok(())
+            }
+            Some(Shard::QueueProxy(proxy)) => {
+                // We should not unproxify a queue proxy shard directly because it can fail if it
+                // fails to send all updates to the remote shard.
+                // Instead we should transform it into a forward proxy shard before unproxify is
+                // called to handle errors at an earlier time.
+                // See `Self::queue_proxy_into_forward_proxy()` for more details.
+                log::warn!(
+                    "Directly unproxifying queue proxy shard, this should not happen normally"
+                );
 
-            // Release max ack version in update handler
-            proxy.set_max_ack_version(None).await;
+                // Finalize, insert local shard back and return finalize result
+                let result = proxy.finalize().await;
+                let (result, local_shard) = match result {
+                    Ok((local_shard, _)) => (Ok(()), local_shard),
+                    Err((err, queue_proxy)) => {
+                        log::error!("Failed to un-proxify local shard because transferring remaining queue items to remote failed: {err}");
+                        let (wrapped_shard, _remote_shard) =
+                            queue_proxy.forget_updates_and_finalize().await;
+                        (Err(err), wrapped_shard)
+                    }
+                };
+                let _ = local_write.insert(Shard::Local(local_shard));
+                result
+            }
+            _ => unreachable!(),
+        }
+    }
 
-            // TODO: also switch state of remote here?
+    /// Revert usage of a `QueueProxy` shard and forget all updates, then un-proxify to local
+    ///
+    /// This can be used to intentionally forget all updates that are collected by the queue proxy
+    /// shard and revert back to a local shard. This is useful if a shard transfer operation using
+    /// a queue proxy must be aborted.
+    ///
+    /// Does nothing if the local shard is not a queue proxy shard.
+    /// This method cannot fail.
+    ///
+    /// # Warning
+    ///
+    /// This intentionally forgets and drops updates pending to be transferred to the remote shard.
+    /// The remote shard may therefore therefore be left in an inconsistent state, which should be
+    /// resolved separately.
+    pub async fn revert_queue_proxy_local(&self) {
+        let mut local_write = self.local.write().await;
 
-            let local_shard = proxy.wrapped_shard;
-            let _ = local_write.insert(Shard::Local(local_shard));
-        }
+        // Take out queue proxy shard or return
+        if !matches!(*local_write, Some(Shard::QueueProxy(_))) {
+            return;
+        };
+        let Some(Shard::QueueProxy(queue_proxy)) = local_write.take() else {
+            unreachable!();
+        };
```

Comment:
```suggestion
        let queue_proxy = match local_write.take() {
            Some(Shard::QueueProxy(queue_proxy)) => queue_proxy,
            shard => {
                *local_write = shard;
                return;
            }
        };
```

Reply:
I had this before. But I don't like taking it out and having to put it back in again. I've made tricky mistakes before using an approach like this. I agree that it doesn't look very nice, but I prefer leaving it in the option if possible.